### ISSUE TITLE
kernel: 4.2.x infinite loop with bond interfaces and bridge fdb command

### ISSUE THREAD

**Issue by kayrus** [coreos/bugs#980]
/cc @dtatulea

I've discovered that the issue appeared in linux kernel 4.2.x. It's caused flannel OOM issue here https://github.com/coreos/flannel/issues/367 which was already fixed in flannel but not in kernel.

It is possible to reproduce the issue by running this script: https://github.com/kayrus/scripts/blob/master/deploy_ubuntu_cluster.sh (ssh ubuntu@ubuntu1 with password: passw0rd)

just run `bridge fdb` and it will run forever

probably problem is somewhere [here](https://github.com/torvalds/linux/blob/v4.2/net/core/rtnetlink.c#L2839)

**Comment by dtatulea** [comment]
There's a very easy way to reproduce this:

```bash
#!/bin/bash

set -x

modprobe bonding
modprobe dummy numdummies=2

echo "+bond0" >  /sys/class/net/bonding_masters 
echo "+dummy0" > /sys/class/net/bond0/bonding/slaves
echo "+dummy1" > /sys/class/net/bond0/bonding/slaves

bridge fdb
```

**Comment by dtatulea** [comment]
TL;DR
An error gets misinterpreted as an index from the switchdev ops (used by the bonding driver) to the rtnetlink fdb dump.

Now the details:

It looks like in 4.2 the bonding driver started using [fdb ops] (https://github.com/torvalds/linux/blob/v4.2/drivers/net/bonding/bond_main.c#L4089) from switchdev which returns [EOPTNOTSUPP] (https://github.com/torvalds/linux/blob/v4.2/net/switchdev/switchdev.c#L317).

This error value gets propagated to the main [fdb dump function](https://github.com/torvalds/linux/blob/v4.2/net/core/rtnetlink.c#L2897) as the _idx_ value which is not expected to be negative and which is forwarded to netlink.

On a 4.1 kernel _idx_ is always > 0.

This code also changed from 4.2 to upstream tip. Looking into how this could be fixed. Not sure if the callbacks should be made to never return an error, or have a check in the [rtnl_fdb_dump](https://github.com/torvalds/linux/blob/1b1050cdc5cdde43177b375b5f22dc070d45d8f8/net/core/rtnetlink.c#L2884) for negative values before assigning to _idx_.

**Comment by dtatulea** [comment]
It looks like the main issue has been [fixed] (https://github.com/torvalds/linux/commit/0890cf6cb6ab1af650025670b1a839671a9a3fcb) in 4.3. 

*However*, that works only if you have CONFIG_NET_SWITCHDEV turned on. Oterwise, you'll still get an error from the unimplemented [switchdev_port_obj_dump](https://github.com/torvalds/linux/blob/1b1050cdc5cdde43177b375b5f22dc070d45d8f8/include/net/switchdev.h#L245)

**Comment by dtatulea** [comment]
Fix was applied to the [net-next](http://marc.info/?t=144766761600004&r=1&w=2) kernel.

**Comment by vcaputo** [comment]
@dtatulea thank you, sir!  I've pulled both the patches in https://github.com/coreos/coreos-overlay/pull/1640, just in case we enable CONFIG_NET_SWITCHDEV.
